python_modules/dagster/dagster/cli/dynamic_loader.py

```diff
@@ -208,68 +208,76 @@ def create_pipeline_loading_mode_data(info):
         raise InvalidPipelineLoadingComboError()
 
 
-DynamicObject = namedtuple('DynamicObject', 'module fn module_name fn_name object')
+class DynamicObject:
+    def __init__(self, module, module_name, fn_name):
+        self.module = module
+        self.module_name = module_name
+        self.fn_name = fn_name
+        self.after_load_fn = None
+        self.object = None
+        self.loaded = False
+
+    def load(self):
+        if self.loaded:
+            module = importlib.reload(self.module)
+        self.loaded = True
+
+        fn = getattr(self.module, self.fn_name)
+        check.is_callable(fn)
+        self.object = fn()
+        if self.after_load_fn:
+            self.object = self.after_load_fn(self.object)
+        return self.object
+
 
 
 def load_file_target_function(file_target_function):
     check.inst_param(file_target_function, 'file_target_function', FileTargetFunction)
     module_name = os.path.splitext(os.path.basename(file_target_function.python_file))[0]
     module = imp.load_source(module_name, file_target_function.python_file)
-    fn = getattr(module, file_target_function.fn_name)
-    check.is_callable(fn)
-    return DynamicObject(module, fn, module_name, file_target_function.fn_name, object=fn())
+    return DynamicObject(module, module_name, file_target_function.fn_name)
 
 
 def load_module_target_function(module_target_function):
     check.inst_param(module_target_function, 'module_target_function', ModuleTargetFunction)
     module = importlib.import_module(module_target_function.module_name)
-    fn = getattr(module, module_target_function.fn_name)
-    check.is_callable(fn)
     return DynamicObject(
-        module, fn, module_target_function.module_name, module_target_function.fn_name, object=fn()
+        module, module_target_function.module_name, module_target_function.fn_name
     )
 
 
 EMPHERMAL_NAME = '<<unnamed>>'
 
 
+def coerce_to_pipeline_to_repository(obj):
+    if not isinstance(obj, RepositoryDefinition):
```

Comment:
let's instead affirmatively check for PipelineDefinition and then throw an error if it is not a Pipeline or a Repository